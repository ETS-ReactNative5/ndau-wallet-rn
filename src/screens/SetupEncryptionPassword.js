import React, { Component } from 'react'
import { StyleSheet, View, ScrollView, Text, TextInput, Alert } from 'react-native'
import CheckBox from 'react-native-check-box'
import FontAwesome from 'react-native-vector-icons/FontAwesome'
import CommonButton from '../components/CommonButton'
import SetupProgressBar from '../components/SetupProgressBar'
import cssStyles from '../css/styles'
import SetupStore from '../model/SetupStore'
import { SafeAreaView } from 'react-navigation'
import MultiSafeHelper from '../helpers/MultiSafeHelper'
import AppConstants from '../AppConstants'

class SetupEncryptionPassword extends Component {
  static PASSWORD_RESET_MODE = 'password-reset'
  static NEW_PASSWORD_MODE = 'new-password'

  constructor (props) {
    super(props)

    this.NEW_PASSWORD_MODE_TEXT =
      'Set a password. This password applies to this ' +
      'app only. Ndau will not have access to it. We cannot help you reset this ' +
      'password, so you should write it down.'
    this.PASSWORD_RESET_MODE_TEXT = 'Please select a new password.'

    this.state = {
      password: '',
      confirmPassword: '',
      showPasswords: false,
      progress: false,
      textInputColor: '#000000',
      mode: AppConstants.NEW_PASSWORD_MODE,
      instructionText: this.NEW_PASSWORD_MODE_TEXT
    }
  }

  componentWillMount () {
    const mode = this.props.navigation.getParam('mode', AppConstants.NEW_PASSWORD_MODE)
    let instructionText = this.NEW_PASSWORD_MODE_TEXT
    if (mode === AppConstants.PASSWORD_RESET_MODE) {
      instructionText = this.PASSWORD_RESET_MODE_TEXT
    }

    this.setState({ mode, instructionText })
  }

  usePassword (event) {
    this.onPushAnother(event)
  }

  checkPasswordsExist = () => {
    return this.state.password === this.state.confirmPassword
  }

  checkPasswordsLength = () => {
    return this.state.password.length >= 8 && this.state.confirmPassword.length >= 8
  }

  showNextSetup = async () => {
    if (!this.checkPasswordsExist()) {
      Alert.alert(
        'Error',
        'The passwords entered do not match.',
        [{ text: 'OK', onPress: () => {} }],
        { cancelable: false }
      )
      this.setState({ textInputColor: '#ff0000' })
      return
    }
    if (!this.checkPasswordsLength()) {
      Alert.alert(
        'Error',
        'The password must be at least 8 characters',
        [{ text: 'OK', onPress: () => {} }],
        { cancelable: false }
      )
      this.setState({ textInputColor: '#ff0000' })
    }

    switch (this.state.mode) {
      case AppConstants.NEW_PASSWORD_MODE: {
        this.finishSetup()
        break
      }
      case AppConstants.PASSWORD_RESET_MODE: {
        await this.resetPassword()
        break
      }
    }
  }

  finishSetup = () => {
    SetupStore.encryptionPassword = this.state.password
    const user = this.props.navigation.getParam('user', null)
    this.props.navigation.navigate('SetupTermsOfService', {
      user,
      walletSetupType: this.props.navigation.state.params &&
        this.props.navigation.state.params.walletSetupType
    })
  }

  resetPassword = async () => {
    const recoveryPhraseString = this.props.navigation.getParam('recoveryPhraseString', null)
    await MultiSafeHelper.resetPassword(recoveryPhraseString, this.state.password)
    const user = await MultiSafeHelper.getDefaultUser(recoveryPhraseString)
    this.props.navigation.navigate('Dashboard', {
      user,
      encryptionPassword: this.state.password,
      walletSetupType: null
    })
  }

  checkedShowPasswords = () => {
    this.setState({ showPasswords: !this.state.showPasswords })
  }

  checkedProgress = () => {
    this.setState({ progress: !this.state.progress })
  }

  showInformation = () => {
    Alert.alert(
      'Information',
      'We use encryption to protect your data. This password protects ' +
        'this app on your mobile only. This is not the same thing as your ' +
        'recovery phrase, which codes for the key to your wallet. We ' +
        'recommend you use a strong password which you do not use anywhere else.',
      [{ text: 'OK', onPress: () => {} }],
      { cancelable: false }
    )
  }

  render () {
    const { textInputColor } = this.state
    return (
      <SafeAreaView style={styles.safeContainer}>
        <View style={cssStyles.container}>
          <ScrollView style={styles.contentContainer}>
            {/* <SetupProgressBar {...this.props} screenNumber={4} /> */}
            <View style={styles.textContainer}>
              <Text style={cssStyles.wizardText} onPress={this.showInformation}>
                {this.state.instructionText}{'  '}
                <FontAwesome name='info' color='#ffffff' size={20} style={{ marginBottom: 3 }} />
              </Text>
            </View>
            <TextInput
              style={cssStyles.textInput}
              onChangeText={password => this.setState({ password })}
              value={this.state.password}
              placeholder='Enter a password'
              placeholderTextColor='#333'
              secureTextEntry={!this.state.showPasswords}
              autoCapitalize='none'
            />
            <TextInput
              style={cssStyles.textInput}
              onChangeText={confirmPassword => this.setState({ confirmPassword })}
              value={this.state.confirmPassword}
              placeholder='Confirm your password'
              placeholderTextColor='#333'
              secureTextEntry={!this.state.showPasswords}
              autoCapitalize='none'
            />
            <View>
              <CheckBox
                style={cssStyles.checkbox}
                onClick={() => this.checkedShowPasswords()}
                isChecked={this.state.showPasswords}
                rightText='Show passwords'
                rightTextStyle={{
                  color: '#ffffff',
                  fontSize: 20,
                  fontFamily: 'TitilliumWeb-Regular'
                }}
                checkBoxColor='#ffffff'
              />
            </View>
            <View>
              <CheckBox
                style={cssStyles.checkbox}
                onClick={() => this.checkedProgress()}
                isChecked={this.state.progress}
                rightText='I understand that ndau cannot help me reset my password, and I will
                have to recover my wallet from my recovery phrase if I lose my password.'
                rightTextStyle={{
                  color: '#ffffff',
                  fontSize: 20,
                  fontFamily: 'TitilliumWeb-Regular'
                }}
                checkBoxColor='#ffffff'
              />
            </View>
          </ScrollView>
          <View style={styles.footer}>
            <CommonButton
              onPress={this.showNextSetup}
              title='Next'
              disabled={!this.state.progress}
            />
          </View>
        </View>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  safeContainer: {
    flex: 1,
    backgroundColor: '#1c2227'
  },
  button: {
    marginTop: 0
  },
  textContainer: {
    marginBottom: 8
  },
  contentContainer: {
    flex: 1 // pushes the footer to the end of the screen
  },
  footer: {
    justifyContent: 'flex-end'
  },
  progress: {
    paddingTop: 15,
    paddingBottom: 15
  },
  infoParagraph: {
    flexDirection: 'row'
  }
})

export default SetupEncryptionPassword
