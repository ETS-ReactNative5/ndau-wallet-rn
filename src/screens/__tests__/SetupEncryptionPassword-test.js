import { mount } from 'enzyme';
import React from 'react';
import { StyleSheet } from 'react-native';
import SetupEncryptionPassword from '../SetupEncryptionPassword';

import renderer from 'react-test-renderer';

describe('testing SetupEncryptionPassword...', () => {
  let styles = StyleSheet.create({
    wizardText: {
      color: '#ffffff',
      fontSize: 20
    }
  });
  const navigator = {
    setStyle: () => {}
  };

  beforeEach(() => {});
  it('renders correctly', () => {
    const tree = renderer
      .create(<SetupEncryptionPassword navigator={navigator} parentStyles={styles} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render parentStyles', () => {
    const wrapper = mount(<SetupEncryptionPassword navigator={navigator} parentStyles={styles} />);
    expect(wrapper.props('parentStyles')).toBeDefined();
  });

  it('throws an error if click without password', () => {
    const wrapper = mount(<SetupEncryptionPassword navigator={navigator} parentStyles={styles} />);
    const onlyButton = wrapper.find('Button').at(0);
    onlyButton.simulate('click');
    const alert = wrapper.find('Alert').at(0);
    expect(alert).toBeDefined();
  });

  it('throws an error if click without confirm password', () => {
    const wrapper = mount(<SetupEncryptionPassword navigator={navigator} parentStyles={styles} />);
    const onlyButton = wrapper.find('Button').at(0);

    wrapper.find('TextInput').at(0).instance().value = 'foo';

    onlyButton.simulate('click');
    const alert = wrapper.find('Alert').at(0);
    expect(alert).toBeDefined();
  });

  it('you can proceed after both have been populated', () => {
    const wrapper = mount(<SetupEncryptionPassword navigator={navigator} parentStyles={styles} />);
    const onlyButton = wrapper.find('Button').at(0);

    wrapper.find('TextInput').at(0).instance().value = 'foo';
    wrapper.find('TextInput').at(1).instance().value = 'foo';
    wrapper.find('CheckBox').at(0).instance().value = true;

    onlyButton.simulate('click');
    const alert = wrapper.find('Alert').at(0);
    expect(alert);
  });
});
