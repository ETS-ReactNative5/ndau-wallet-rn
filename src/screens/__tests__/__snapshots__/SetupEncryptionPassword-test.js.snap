// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SetupEncryptionPassword... renders correctly 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 26.5,
        "paddingLeft": 15,
        "paddingRight": 15,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <Image
          resizeMode="contain"
          screenNumber={4}
          source={
            Object {
              "testUri": "../../../img/step-4.png",
            }
          }
          style={
            Object {
              "backgroundColor": "transparent",
              "height": 60,
              "marginBottom": 5,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "marginBottom": 8,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            onPress={[Function]}
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 20,
              }
            }
          >
            Set a password. This password applies to this app only. Ndau will not have access to it. We cannot help you reset this password, so you should write it down.
              
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 20,
                  },
                  Object {
                    "marginBottom": 3,
                  },
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              ï„©
            </Text>
          </Text>
        </View>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="Enter a password"
          placeholderTextColor="#333"
          secureTextEntry={true}
          style={
            Object {
              "backgroundColor": "#ffffff",
              "borderColor": "gray",
              "borderRadius": 6,
              "borderWidth": 1,
              "color": "#000000",
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 18,
              "height": 93.5,
              "marginTop": 13.5,
              "paddingLeft": 7.5,
            }
          }
          value=""
        />
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="Confirm your password"
          placeholderTextColor="#333"
          secureTextEntry={true}
          style={
            Object {
              "backgroundColor": "#ffffff",
              "borderColor": "gray",
              "borderRadius": 6,
              "borderWidth": 1,
              "color": "#000000",
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 18,
              "height": 93.5,
              "marginTop": 13.5,
              "paddingLeft": 7.5,
            }
          }
          value=""
        />
        <View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "paddingBottom": 13.5,
                "paddingTop": 13.5,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Image
                source={
                  Object {
                    "testUri": "../../react-native-check-box/img/ic_check_box_outline_blank.png",
                  }
                }
                style={
                  Object {
                    "tintColor": "#ffffff",
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "marginLeft": 10,
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "TitilliumWeb-Regular",
                      "fontSize": 20,
                    },
                  ]
                }
              >
                Show passwords
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "paddingBottom": 13.5,
                "paddingTop": 13.5,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Image
                source={
                  Object {
                    "testUri": "../../react-native-check-box/img/ic_check_box_outline_blank.png",
                  }
                }
                style={
                  Object {
                    "tintColor": "#ffffff",
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "marginLeft": 10,
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "TitilliumWeb-Regular",
                      "fontSize": 20,
                    },
                  ]
                }
              >
                I understand that ndau cannot help me reset my password, and I will have to recover my wallet from my recovery phrase if I lose my password.
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignContent": "center",
            "backgroundColor": "#696969",
            "borderRadius": 3,
            "height": 93.5,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007aff",
                "fontSize": 17,
                "fontWeight": "500",
                "textAlign": "center",
              },
              Object {
                "color": "#dcdcdc",
              },
              Object {
                "color": "#fff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 16,
              },
              undefined,
            ]
          }
        >
          Next
        </Text>
      </View>
    </View>
  </View>
</View>
`;
